### ISSUE TITLE
Body.updateAABB increases CPU usage over time

### ISSUE THREAD

**Issue by cnatale** [schteppe/p2.js#177]
Test browser: Chrome 43.0.2357.134 (64-bit) Mac

I am working with a modified version of the ragdoll demo (you can view source here: https://bitbucket.org/chrisnatale/ihtai/src/d0aedb966d2f2011c7ffa33d15dd73effc7272aa/ragdolldemo.html?at=master). The total number of objects onscreen is static over time. There are four planes that serve as a container for the ragdoll, and the ragdoll itself.

One thing I am noticing while profiling is that total CPU usage by the Body.updateAABB() method continually increases over time. If I leave the simulation running overnight, performance slows to a crawl due to CPU usage of this method. Percentage of total CPU usage for this method increases as follows:

5 minutes: 1.57%
10 minutes: 2.93%
15 minutes: 5.14%
~8 hours: 53.84%

Looking at the method and methods calling it, nothing immediately jumps out. Any ideas on what would be causing this?

Attached: screenshots of Chrome's CPU Usage profiler results.

Five Minutes:
<img width="1144" alt="five minutes" src="https://cloud.githubusercontent.com/assets/2414716/8766942/1217c0e2-2e00-11e5-907f-a7a058330210.png">

Ten Minutes:
<img width="1144" alt="ten minutes" src="https://cloud.githubusercontent.com/assets/2414716/8766945/121babe4-2e00-11e5-8246-a4a5476db6c1.png">

Fifteen Minutes:
<img width="1145" alt="fifteen minutes" src="https://cloud.githubusercontent.com/assets/2414716/8766943/121a35c0-2e00-11e5-9ddc-43f50aa6012d.png">

Eight Hours:
<img width="1145" alt="eight hours" src="https://cloud.githubusercontent.com/assets/2414716/8766944/121ab360-2e00-11e5-8c3a-ef8222567e46.png">

Update: here is the expanded call stack from the simulation after it has been running for ~eight hours:
<img width="1142" alt="eight hours expanded callstack" src="https://cloud.githubusercontent.com/assets/2414716/8767159/a135d644-2e08-11e5-8167-bd5c8b600751.png">

**Comment by schteppe** [comment]
Hi,
Very interesting.. Actually I have no idea what would cause this. Memory leak? Will have a look whenever I get some time over.

**Comment by cnatale** [comment]
Memory appears to be holding steady in these tests. I'll take some screen caps the next time I run it overnight, but I observed it holding at ~257 megs.

**Comment by cnatale** [comment]
I ran the same memory and CPU usage tests on the ragdoll demo from P2's GitHub page overnight. There is no comparable increase in CPU usage of the Body.updateAABB() method over time.

I'm going to run more tests over the next few days to try and isolate which change I made to the demo introduced this problem.

**Comment by cnatale** [comment]
I've isolated the situation where this bug manifests. I'm using the following code to generate four planes that serve as a container for the ragdoll. When I remove the left, top, and right planes, it no longer manifests.

                var OTHER =     Math.pow(2,1),
                    BODYPARTS = Math.pow(2,2),
                    WALLS =     Math.pow(2,3),
                    bodyPartShapes = []; 

                // ...

                // Create ground
                var groundShape = new p2.Plane();
                var ground = new p2.Body({
                    position:[0,-1],
                });
                ground.addShape(groundShape);
                groundShape.collisionGroup = WALLS;
                groundShape.collisionMask =  BODYPARTS|OTHER;
                world.addBody(ground);

                //create walls
                var leftWallShape = new p2.Plane();
                var leftWall = new p2.Body({
                    position:[-8,0],
                    angle:[3*Math.PI/2]
                });

                leftWall.addShape(leftWallShape);
                leftWallShape.collisionGroup = WALLS;
                leftWallShape.collisionMask =  BODYPARTS|OTHER;
                world.addBody(leftWall);            

                var rightWallShape = new p2.Plane();
                var rightWall = new p2.Body({
                    position:[8,0],
                    angle:[Math.PI/2]
                });

                rightWall.addShape(rightWallShape);
                rightWallShape.collisionGroup = WALLS;
                rightWallShape.collisionMask =  BODYPARTS|OTHER;    
                world.addBody(rightWall);               

                var topWallShape = new p2.Plane();
                var topWall = new p2.Body({
                    position:[0,10],
                    angle:[Math.PI]
                });

                topWall.addShape(topWallShape);
                topWallShape.collisionGroup = WALLS;
                topWallShape.collisionMask =  BODYPARTS|OTHER;                  
                world.addBody(topWall);

**Comment by schteppe** [comment]
Thanks for your help. I've tried to reproduce your test case but I'm not sure if I'm succeeding... So I optimized away some suspicious code. Can you try this patch? (just copy/paste it in any script after p2.js was included)

```js
p2.Plane.prototype.computeAABB = function(out, position, angle){
    var a = angle % (2 * Math.PI);
    var set = p2.vec2.set;
    var max = 1e7;
    var lowerBound = out.lowerBound;
    var upperBound = out.upperBound;

    // Set max bounds
    set(lowerBound, -max, -max);
    set(upperBound,  max,  max);

    if(a === 0){
        // y goes from -inf to 0
        upperBound[1] = 0;

    } else if(a === Math.PI / 2){

        // x goes from 0 to inf
        lowerBound[0] = 0;

    } else if(a === Math.PI){

        // y goes from 0 to inf
        lowerBound[1] = 0;

    } else if(a === 3*Math.PI/2){

        // x goes from -inf to 0
        upperBound[0] = 0;

    }
};
```

Not sure if this is going to help, but might be worth a try.

**Comment by cnatale** [comment]
Thanks. I'm still seeing a gradual increase in CPU usage over time after adding the prototype method. When I get some time, I'll step through the source at various intervals and see if I can spot what exactly is causing updateAABB() to either take longer to run or be triggered much more often the longer the demo is running.

**Comment by cnatale** [comment]
I've narrowed this down to the garbage collector getting triggered almost every time P2's rendering loop executes. I've included a screen shot of a few of the cycles. This fits with not actually seeing the total memory climb, yet performance slowing to a crawl. Basically, it looks like the library is cleaning up after itself, but allocating a lot of memory that leads to the constant GC calls. I still have no idea where the ~14 megs of allocations every cycle are coming from, though.

<img width="1270" alt="screen shot 2015-07-29 at 7 20 27 pm" src="https://cloud.githubusercontent.com/assets/2414716/8974529/766bc87a-3627-11e5-9d37-23054deae20e.png">

**Comment by cnatale** [comment]
I ran another test with the goal of isolating where the problem is coming from. I ripped out the WebGLRenderer and used a custom-rolled Canvas renderer based on P2's Canvas demos. I turned off my a.i. engine and applied random forces to limbs every cycle. The results were the same as before. Performance degrading slowly over time, until it slows to a crawl after about an hour of simulation. This suggests that it is indeed a problem with the P2 core library.

**Comment by cnatale** [comment]
I think I finally figured out what is going on. I turned down the number of substeps to 1, and this fixed the problem. I'm hypothesizing that Chrome's gc is unable to keep up with the flow of released variables when the number of substeps is set to the default 10 per iteration. In particular, the beginContact and endContact events being emitted are churning out a ton of Objects (see screenshot below). 

To recap, this appeared to be caused by updateAABB() and a few other methods in P2's update cycle because the gc was getting triggered during these methods, making them appear slower than they actually are. Sorry for the false lead.

I tried commenting out where these events were emitted in the P2 source, but this only slowed the performance degredation. This suggests that there's no magic bullet optimization to the source to prevent this. Since decreasing substeps is an acceptable solution for me, I consider this bug closed.
 
<img width="1062" alt="screen shot 2015-08-01 at 12 44 10 pm" src="https://cloud.githubusercontent.com/assets/2414716/9024392/3676bed4-3882-11e5-8713-2ea65d8d8d9b.png">
